File: etl.py
"""
ETL layer — transform raw PMS API JSON into domain dataclasses.

Responsibilities:
  * Dynamic field detection on the first sample reservation.
  * Category-ID → name mapping (including Mews localised ``Names`` dicts).
  * Building a reservation-to-rate lookup from the ``Items`` array when
    rates are not embedded in the reservation itself.
  * Parsing each raw dict into a ``Reservation`` (with ``booking_window``).
  * Extracting ``RoomType`` records for database storage.
"""

from __future__ import annotations

from typing import Any, Optional

from config import logger
from models import FieldMap, Reservation, RoomType
from utils import detect_field, get_nested, parse_date


# ── Field detection ───────────────────────────────────────────────────────────

def detect_field_map(sample: dict[str, Any]) -> FieldMap:
    """Inspect a sample reservation and auto-detect API field names."""
    return FieldMap(
        reservation_id=detect_field(
            sample, ["Id", "ReservationId", "id"], "reservation_id",
        ),
        stay_date=detect_field(
            sample,
            ["ArrivalDate", "ScheduledStartUtc", "StartUtc", "CheckInUtc"],
            "stay_date",
        ),
        booking_date=detect_field(
            sample,
            ["CreatedUtc", "Created", "BookingDate", "CreateDate"],
            "booking_date",
        ),
        room_type_id=detect_field(
            sample,
            [
                "SpaceCategoryId",
                "RoomCategoryId",
                "ResourceCategoryId",
                "RequestedCategoryId",
            ],
            "room_type_id",
        ),
        rate=detect_field(
            sample,
            [
                "StayPriceIncludingTaxes",
                "TotalAmount.Value",
                "Rate.Value",
                "Price.Value",
                "TotalPrice",
                "Amount",
            ],
            "rate",
        ),
    )


# ── Category helpers ──────────────────────────────────────────────────────────

_PREFERRED_LOCALES = ("en-US", "en-GB", "en-AU", "en")


def _resolve_name(cat: dict[str, Any]) -> str:
    """Extract a human-readable name from a category dict.

    Mews API ≥2024 uses a localised ``Names`` dict instead of a flat
    ``Name`` string.  We prefer ``en-US`` → ``en-GB`` → first available.
    Falls back to ``ShortName`` or ``"Unknown"``.
    """
    # flat Name field (older or simpler APIs)
    flat = cat.get("Name")
    if isinstance(flat, str) and flat:
        return flat

    # localised Names dict
    names: dict[str, str] | None = cat.get("Names")
    if isinstance(names, dict) and names:
        for locale in _PREFERRED_LOCALES:
            if locale in names:
                return names[locale]
        # take the first available locale
        return next(iter(names.values()))

    # ShortName / ShortNames fallback
    short = cat.get("ShortName")
    if isinstance(short, str) and short:
        return short
    short_names: dict[str, str] | None = cat.get("ShortNames")
    if isinstance(short_names, dict) and short_names:
        for locale in _PREFERRED_LOCALES:
            if locale in short_names:
                return short_names[locale]
        return next(iter(short_names.values()))

    return "Unknown"


def build_category_map(data: dict[str, Any]) -> dict[str, str]:
    """Build a mapping of category-ID → category-name from the API response."""
    categories: dict[str, str] = {}
    for key in ("SpaceCategories", "RoomCategories", "ResourceCategories"):
        items = data.get(key)
        if not isinstance(items, list):
            continue
        for cat in items:
            cat_id = cat.get("Id") or cat.get("id")
            cat_name = _resolve_name(cat)
            if cat_id:
                categories[cat_id] = cat_name
        break
    return categories


def extract_room_types(
    categories: dict[str, str],
    hotel_id: int,
) -> list[RoomType]:
    """Convert the category map into ``RoomType`` instances."""
    return [
        RoomType(hotel_id=hotel_id, category_id=cid, name=cname)
        for cid, cname in categories.items()
    ]


# ── Item-based rate lookup ───────────────────────────────────────────────────

def build_rate_lookup(data: dict[str, Any]) -> dict[str, float]:
    """Build a reservation-ID → total-rate mapping from the Items array.

    The Mews Connector API returns line-item charges in an ``Items``
    (or ``OrderItems``) array.  Each item's ``OrderId`` matches the
    reservation ``Id``.  We sum ``Amount.GrossValue`` (falling back
    to ``Amount.Value``) per reservation to get the total stay rate.
    """
    rate_map: dict[str, float] = {}
    items: list[dict[str, Any]] | None = data.get("Items") or data.get("OrderItems")
    if not isinstance(items, list):
        return rate_map

    for item in items:
        order_id = item.get("OrderId")
        if not order_id:
            continue
        amount = item.get("Amount")
        if not isinstance(amount, dict):
            continue
        val = amount.get("GrossValue") or amount.get("Value")
        if val is None:
            continue
        try:
            rate_map[order_id] = rate_map.get(order_id, 0.0) + float(val)
        except (ValueError, TypeError):
            pass

    return rate_map


# ── Reservation list discovery ────────────────────────────────────────────────

def find_reservations_list(data: dict[str, Any]) -> list[dict[str, Any]]:
    """Locate the reservations array in the API response."""
    for key in ("Reservations", "reservations"):
        items = data.get(key)
        if isinstance(items, list) and len(items) > 0:
            return items
    # Only fall back to Items if there is no Reservations key at all
    for key in ("Items", "items"):
        items = data.get(key)
        if isinstance(items, list) and len(items) > 0:
            return items
    return []


# ── Rate extraction ───────────────────────────────────────────────────────────

def extract_rate(raw: dict[str, Any], rate_field: Optional[str]) -> Optional[float]:
    """Extract and normalise the rate value, handling nested dicts."""
    val = get_nested(raw, rate_field)
    if isinstance(val, dict):
        val = val.get("Value")
    if val is None:
        return None
    try:
        return float(val)
    except (ValueError, TypeError):
        return None


# ── Single-reservation parsing ────────────────────────────────────────────────

def parse_single_reservation(
    raw: dict[str, Any],
    field_map: FieldMap,
    categories: dict[str, str],
    hotel_id: int,
    rate_lookup: dict[str, float] | None = None,
) -> Reservation:
    """Transform one raw API dict into a ``Reservation`` dataclass."""
    stay_dt = parse_date(get_nested(raw, field_map.stay_date))
    book_dt = parse_date(get_nested(raw, field_map.booking_date))
    rt_id = get_nested(raw, field_map.room_type_id)

    window: Optional[int] = None
    if stay_dt and book_dt:
        window = (stay_dt - book_dt).days

    # Try the inline rate field first, then fall back to the Items-based lookup
    rate = extract_rate(raw, field_map.rate)
    if rate is None and rate_lookup:
        res_id = get_nested(raw, field_map.reservation_id)
        if res_id:
            rate = rate_lookup.get(res_id)

    return Reservation(
        hotel_id=hotel_id,
        reservation_id=get_nested(raw, field_map.reservation_id),
        room_type=categories.get(rt_id, rt_id or "Unknown"),
        stay_date=stay_dt,
        booking_date=book_dt,
        booking_window=window,
        rate=rate,
        base_rate=rate,
        raw_json=raw,
    )


# ── Batch parsing ─────────────────────────────────────────────────────────────

def parse_api_response(
    data: dict[str, Any],
    hotel_id: int,
) -> tuple[list[Reservation], list[RoomType]]:
    """Full ETL pass: detect fields → parse all reservations + room types."""
    raw_list = find_reservations_list(data)
    if not raw_list:
        logger.warning("Hotel %d: API response contained no reservations.", hotel_id)
        return [], []

    field_map = detect_field_map(raw_list[0])
    logger.info("Hotel %d: detected fields %s", hotel_id, field_map)

    categories = build_category_map(data)
    rate_lookup = build_rate_lookup(data)
    if rate_lookup:
        logger.info("Hotel %d: built rate lookup from Items (%d entries).",
                     hotel_id, len(rate_lookup))

    reservations = [
        parse_single_reservation(r, field_map, categories, hotel_id, rate_lookup)
        for r in raw_list
    ]
    room_types = extract_room_types(categories, hotel_id)
    return reservations, room_types

File: maya-rms/supabase/functions/_shared/mews/etl.ts
/**
 * Mews → Supabase row shapes. Mirrors `shared/legacy-python/etl.py` (field
 * detection, category names, Items-based rate lookup).
 */

import { addCalendarDays } from "../engine/timezone.ts";
import { redactMewsPayload } from "../pms/redact.ts";

type Json = Record<string, unknown>;

export type FieldMap = {
  reservation_id: string | null;
  stay_date: string | null;
  booking_date: string | null;
  room_type_id: string | null;
  rate: string | null;
};

const PREFERRED_LOCALES = ["en-US", "en-GB", "en-AU", "en"] as const;

export function getNested(obj: unknown, dottedKey: string | null): unknown {
  if (dottedKey === null || typeof obj !== "object" || obj === null) return undefined;
  let val: unknown = obj;
  for (const part of dottedKey.split(".")) {
    if (typeof val !== "object" || val === null) return undefined;
    val = (val as Json)[part];
  }
  return val;
}

export function detectField(sample: Json, candidates: string[]): string | null {
  for (const candidate of candidates) {
    if (getNested(sample, candidate) != null) return candidate;
  }
  return null;
}

export function detectFieldMap(sample: Json): FieldMap {
  return {
    reservation_id: detectField(sample, ["Id", "ReservationId", "id"]),
    stay_date: detectField(sample, [
      "ArrivalDate",
      "ScheduledStartUtc",
      "StartUtc",
      "CheckInUtc",
    ]),
    booking_date: detectField(sample, [
      "CreatedUtc",
      "Created",
      "BookingDate",
      "CreateDate",
    ]),
    room_type_id: detectField(sample, [
      "SpaceCategoryId",
      "RoomCategoryId",
      "ResourceCategoryId",
      "RequestedCategoryId",
    ]),
    rate: detectField(sample, [
      "StayPriceIncludingTaxes",
      "TotalAmount.Value",
      "Rate.Value",
      "Price.Value",
      "TotalPrice",
      "Amount",
    ]),
  };
}

function detectDepartureField(sample: Json): string | null {
  return detectField(sample, ["DepartureDate", "ScheduledEndUtc", "EndUtc", "CheckOutUtc"]);
}

function resolveCategoryName(cat: Json): string {
  const flat = cat.Name;
  if (typeof flat === "string" && flat) return flat;

  const names = cat.Names;
  if (names && typeof names === "object" && names !== null) {
    const dict = names as Record<string, string>;
    for (const loc of PREFERRED_LOCALES) {
      if (dict[loc]) return dict[loc];
    }
    const first = Object.values(dict)[0];
    if (first) return first;
  }

  const short = cat.ShortName;
  if (typeof short === "string" && short) return short;

  const shortNames = cat.ShortNames;
  if (shortNames && typeof shortNames === "object" && shortNames !== null) {
    const dict = shortNames as Record<string, string>;
    for (const loc of PREFERRED_LOCALES) {
      if (dict[loc]) return dict[loc];
    }
    const first = Object.values(dict)[0];
    if (first) return first;
  }

  return "Unknown";
}

export type CategoryInventoryEntry = { name: string; total_rooms: number };

/** Prefer explicit inventory fields from the PMS category payload; else use hotel default. */
function resolveCategoryRoomCount(cat: Json, fallback: number): number {
  const candidates: unknown[] = [
    cat.RoomCount,
    cat.SpacesCount,
    cat.ResourceCount,
    cat.InventoryCount,
  ];
  for (const raw of candidates) {
    if (typeof raw === "number" && Number.isFinite(raw) && raw > 0) {
      return Math.min(100_000, Math.floor(raw));
    }
    if (typeof raw === "string" && /^\d+$/.test(raw.trim())) {
      const n = parseInt(raw.trim(), 10);
      if (n > 0) return Math.min(100_000, n);
    }
  }
  return fallback;
}

export function buildCategoryInventory(
  data: Json,
  defaultRoomsPerCategory: number,
): Record<string, CategoryInventoryEntry> {
  const categories: Record<string, CategoryInventoryEntry> = {};
  for (const key of ["SpaceCategories", "RoomCategories", "ResourceCategories"] as const) {
    const items = data[key];
    if (!Array.isArray(items)) continue;
    for (const cat of items) {
      if (!cat || typeof cat !== "object") continue;
      const c = cat as Json;
      const id = c.Id ?? c.id;
      if (typeof id === "string" && id) {
        categories[id] = {
          name: resolveCategoryName(c),
          total_rooms: resolveCategoryRoomCount(c, defaultRoomsPerCategory),
        };
      }
    }
    break;
  }
  return categories;
}

export function buildCategoryMap(data: Json): Record<string, string> {
  const inv = buildCategoryInventory(data, 100);
  return Object.fromEntries(Object.entries(inv).map(([k, v]) => [k, v.name]));
}

/** Real Mews Ids are GUIDs, but a numeric one still has to dedupe. */
function itemDedupeKey(item: Json): string | null {
  const id = item.Id;
  if (typeof id === "string") return id || null;
  if (typeof id === "number" && Number.isFinite(id)) return String(id);
  return null;
}

export function buildRateLookup(data: Json): Record<string, number> {
  const rateMap: Record<string, number> = {};
  const items = data.Items ?? data.OrderItems;
  if (!Array.isArray(items)) return rateMap;

  // The sync walks the range in 96h windows and Mews returns any stay that
  // collides with a window — plus that stay's whole order — in every window it
  // touches. The chunks are concatenated, so a stay crossing one boundary
  // arrives with its items twice and would sum to double the stay total.
  const countedItemIds = new Set<string>();

  for (const item of items) {
    if (!item || typeof item !== "object") continue;
    const orderId = (item as Json).OrderId;
    if (typeof orderId !== "string" || !orderId) continue;
    const amount = (item as Json).Amount;
    if (!amount || typeof amount !== "object") continue;
    const a = amount as Json;
    const val = a.GrossValue ?? a.Value;
    if (val === null || val === undefined) continue;
    const n = Number(val);
    if (!Number.isFinite(n)) continue;
    const itemId = itemDedupeKey(item as Json);
    if (itemId !== null) {
      if (countedItemIds.has(itemId)) continue;
      countedItemIds.add(itemId);
    }
    rateMap[orderId] = (rateMap[orderId] ?? 0) + n;
  }
  return rateMap;
}

/** Mews service order state for canceled bookings (legacy + current APIs). */
export function isCanceledMewsReservation(raw: Json): boolean {
  const s = raw.State ?? raw.state;
  if (typeof s !== "string") return false;
  const n = s.trim().toLowerCase();
  return n === "canceled" || n === "cancelled";
}

export function findReservationsList(data: Json): Json[] {
  for (const key of ["Reservations", "reservations"] as const) {
    const items = data[key];
    if (Array.isArray(items) && items.length > 0) return items as Json[];
  }
  for (const key of ["Items", "items"] as const) {
    const items = data[key];
    if (Array.isArray(items) && items.length > 0) return items as Json[];
  }
  return [];
}

function extractRate(raw: Json, rateField: string | null): number | null {
  if (rateField) {
    let val = getNested(raw, rateField);
    if (val && typeof val === "object" && "Value" in (val as Json)) {
      val = (val as Json).Value;
    }
    if (val !== null && val !== undefined) {
      const n = Number(val);
      if (Number.isFinite(n)) return n;
    }
  }
  return null;
}

const ZONE_QUALIFIED = /(?:Z|[+-]\d{2}:?\d{2})$/i;
const LEADING_YMD = /^(\d{4}-\d{2}-\d{2})/;
const YMD_PARTS = { year: "numeric", month: "2-digit", day: "2-digit" } as const;

const formatterByTimeZone = new Map<string, Intl.DateTimeFormat>();

/**
 * One formatter per zone: a fleet sync parses tens of thousands of reservations
 * in a single invocation and building these per row costs more than the rest of
 * the parse put together. A junk `hotels.timezone` makes Intl throw, which would
 * fail the whole hotel instead of one row, so it degrades to UTC.
 */
function hotelDateFormatter(timeZone: string): Intl.DateTimeFormat {
  const cached = formatterByTimeZone.get(timeZone);
  if (cached) return cached;
  let fmt: Intl.DateTimeFormat;
  try {
    fmt = new Intl.DateTimeFormat("en-CA", { timeZone, ...YMD_PARTS });
  } catch {
    fmt = new Intl.DateTimeFormat("en-CA", { timeZone: "UTC", ...YMD_PARTS });
  }
  formatterByTimeZone.set(timeZone, fmt);
  return fmt;
}

/**
 * Mews names its instants `…Utc`, and those already fold in the property's local
 * check-in time — a 5pm Los Angeles arrival is midnight UTC the next day, so
 * truncating in UTC drops that arrival night. `ArrivalDate` and friends name a
 * civil day instead, and reading one of those as an instant shifts a hotel west
 * of Greenwich a day early — the same bug pointing the other way. Zone-less
 * timestamps are civil too: parsing them would use the machine's zone.
 */
function fieldToHotelDate(raw: Json, field: string | null, timeZone: string): string | null {
  if (!field) return null;
  const val = getNested(raw, field);
  if (typeof val !== "string") return null;
  const s = val.trim();
  if (!field.endsWith("Utc") || !ZONE_QUALIFIED.test(s)) {
    return LEADING_YMD.exec(s)?.[1] ?? null;
  }
  const d = new Date(s.replace(/Z$/, "+00:00"));
  if (Number.isNaN(d.getTime())) return null;
  if (timeZone === "UTC") return d.toISOString().slice(0, 10);
  return hotelDateFormatter(timeZone).format(d);
}

/**
 * Inclusive arrival, exclusive departure (hotel-night semantics), on the hotel's
 * own calendar. Same local day for arrival and departure yields a single stay night.
 */
export function enumerateStayNights(
  raw: Json,
  fieldMap: FieldMap,
  hotelTimeZone = "UTC",
): { nights: string[] } {
  const startStr = fieldToHotelDate(raw, fieldMap.stay_date, hotelTimeZone);
  if (!startStr) return { nights: [] };

  const endStr = fieldToHotelDate(raw, detectDepartureField(raw), hotelTimeZone);

  let nights: string[];
  if (!endStr || endStr <= startStr) {
    nights = [startStr];
  } else {
    nights = [];
    let cur = startStr;
    while (cur < endStr) {
      nights.push(cur);
      cur = addCalendarDays(cur, 1);
    }
  }

  return { nights };
}

/**
 * Lead time against the stay night itself, not arrival — night 2 of a stay
 * is one day further out than night 1, and pickup windows compare per-night.
 */
function bookingWindowFor(bookingDate: string, stay: string): number {
  const a = new Date(`${stay}T00:00:00Z`).getTime();
  const b = new Date(`${bookingDate}T00:00:00Z`).getTime();
  return Math.max(0, Math.round((a - b) / 86400000));
}

export type ParsedRoomType = {
  external_room_type_id: string;
  name: string;
  display_name: string | null;
  /** Physical room count for this category; from PMS or `defaultRoomsPerCategory`. */
  total_rooms: number;
};

export type ParsedReservationRow = {
  external_reservation_id: string;
  external_room_type_id: string | null;
  stay_date: string;
  booking_date: string | null;
  booking_window_days: number | null;
  /** Nightly rate from Mews; `base_rate` in DB is set by trigger (original BAR). */
  current_rate: number | null;
  raw_payload: Json;
};

/** Observability for ingest edge cases (empty when there are no reservations in the payload). */
export type ParsedMewsStats = {
  skippedMissingReservationId: number;
  skippedNoStayNights: number;
  duplicateStayNightKeysMerged: number;
  rowsWithMissingRate: number;
  skippedCanceled: number;
};

const emptyStats = (): ParsedMewsStats => ({
  skippedMissingReservationId: 0,
  skippedNoStayNights: 0,
  duplicateStayNightKeysMerged: 0,
  rowsWithMissingRate: 0,
  skippedCanceled: 0,
});

export function parseMewsApiResponse(
  data: Json,
  options?: { defaultRoomsPerCategory?: number; hotelTimeZone?: string | null },
): {
  roomTypes: ParsedRoomType[];
  reservations: ParsedReservationRow[];
  stats: ParsedMewsStats;
  canceledExternalIds: string[];
} {
  const rawList = findReservationsList(data);
  if (rawList.length === 0) {
    return { roomTypes: [], reservations: [], stats: emptyStats(), canceledExternalIds: [] };
  }

  const fieldMap = detectFieldMap(rawList[0]);
  const tz = options?.hotelTimeZone || "UTC";
  const defaultRooms = options?.defaultRoomsPerCategory ?? 100;
  const categories = buildCategoryInventory(data, defaultRooms);
  const rateLookup = buildRateLookup(data);

  const roomTypes: ParsedRoomType[] = Object.entries(categories).map(([id, entry]) => ({
    external_room_type_id: id,
    name: entry.name,
    display_name: entry.name,
    total_rooms: entry.total_rooms,
  }));

  const stats = emptyStats();
  const byStayNight = new Map<string, ParsedReservationRow>();
  const canceledExternalIds = new Set<string>();

  for (const raw of rawList) {
    const rid = fieldMap.reservation_id ? getNested(raw, fieldMap.reservation_id) : null;
    if (typeof rid !== "string" || !rid) {
      stats.skippedMissingReservationId += 1;
      continue;
    }

    if (isCanceledMewsReservation(raw)) {
      canceledExternalIds.add(rid);
      stats.skippedCanceled += 1;
      continue;
    }

    const rtIdRaw = fieldMap.room_type_id ? getNested(raw, fieldMap.room_type_id) : null;
    const externalRoomTypeId = typeof rtIdRaw === "string" ? rtIdRaw : null;

    let rate = extractRate(raw, fieldMap.rate);
    if (rate === null && fieldMap.reservation_id) {
      const idKey = getNested(raw, fieldMap.reservation_id);
      if (typeof idKey === "string" && rateLookup[idKey] !== undefined) {
        rate = rateLookup[idKey];
      }
    }

    const { nights } = enumerateStayNights(raw, fieldMap, tz);
    if (nights.length === 0) {
      stats.skippedNoStayNights += 1;
      continue;
    }

    const denom = Math.max(1, nights.length);
    const nightly = rate !== null ? Math.round((rate / denom) * 100) / 100 : null;

    // Same calendar as the stay nights, or an evening booking would read as
    // one day later than it was and shorten every lead time on the stay.
    const bookingDate = fieldToHotelDate(raw, fieldMap.booking_date, tz);

    for (const night of nights) {
      const key = `${rid}:${night}`;
      const row: ParsedReservationRow = {
        external_reservation_id: rid,
        external_room_type_id: externalRoomTypeId,
        stay_date: night,
        booking_date: bookingDate,
        booking_window_days: bookingDate ? bookingWindowFor(bookingDate, night) : null,
        current_rate: nightly,
        raw_payload: redactMewsPayload(raw),
      };
      if (byStayNight.has(key)) {
        stats.duplicateStayNightKeysMerged += 1;
      }
      byStayNight.set(key, row);
    }
  }

  for (const row of byStayNight.values()) {
    if (row.current_rate === null) {
      stats.rowsWithMissingRate += 1;
    }
  }

  return {
    roomTypes,
    reservations: [...byStayNight.values()],
    stats,
    canceledExternalIds: [...canceledExternalIds],
  };
}
